### PR TITLE
Gradient checkpointing libraries

### PR DESCRIPTION
**Pull request by NikZak**
These are the libraries for #711

This PR does not change the existing files

It contains graph_editor and memory_saving_gradients.py used in #711

### CODE REVIEW

File: efficientdet/graph_editor/BUILD [hunk 1]
```diff
@@ -0,0 +1,162 @@
+# Description:
+#   contains parts of TensorFlow that are experimental or unstable and which are not supported.
+
```
Comment (fsx950223) [inline efficientdet/graph_editor/BUILD:3]: Maybe we could remove BUILD files?

Comment (NikZak) [reply efficientdet/graph_editor/BUILD]: Sure, removed

File: efficientdet/gradient_checkpointing/LICENSE.txt [hunk 2]
```diff
@@ -0,0 +1,21 @@
+The MIT License
+
+Copyright (c) 2018 OpenAI (http://openai.com)
```
Comment (mingxingtan) [inline efficientdet/gradient_checkpointing/LICENSE.txt:3]: Hmm, we are not able to import code with such copyright.

Comment (kartik4949) [reply efficientdet/gradient_checkpointing/LICENSE.txt:3]: @mingxingtan memory_saving_gradients is open source codebase (openai) ,cant we keep the copyright?

Comment (NikZak) [reply efficientdet/gradient_checkpointing/LICENSE.txt:3]: I see. That's too bad. Is there any workaround? In the meantime I may try to ask the authors if there's a way to remove the copyright keeping mentioning their names. How can l I explain why it has to be removed?

Comment (NikZak) [reply efficientdet/gradient_checkpointing/LICENSE.txt:3]: @mingxingtan
What if keep all the libraries as a separate repository? Then you don't need to copy them here, you would be able to use them with "pip install" and import. Is it a valid workaround?

Comment (NikZak) [reply efficientdet/gradient_checkpointing/LICENSE.txt:3]: Great, thanks. Waiting for your resolution then

**Comment by mingxingtan** [comment]
We have decided to merge this third_party code!
@NikZak Could you run yafq and pylint to reformat your code? Thanks!

**Comment by fsx950223** [comment]
How about move it to ```third_party``` directory?

**Comment by NikZak** [comment]
Thanks. I will yapf, pylint and adopt to latest changes

**Comment by mingxingtan** [comment]
Good idea @fsx950223 !

@NikZak Could you create a third_party folder and move this code to third_party? You can slightly simply folder names, such as:

gradient_checkpointing  -> third_party/grad_checkpoint
graph_editor -> third_party/graph_edit

**Comment by NikZak** [comment]
Ok, this is done. Please merge this one

File: efficientdet/third_party/graph_edit/edit.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from third_party.graph_edit import reroute
+from third_party.graph_edit import select
+from third_party.graph_edit import subgraph
+from third_party.graph_edit import util
+from tensorflow.python.ops import array_ops as tf_array_ops
```
Comment (mingxingtan) [inline efficientdet/third_party/graph_edit/edit.py:25]: It would be better to avoid importing tf internal stuffs?

For example, here you can simply import tensorflow as tf, and replace the tf_array_ops.placeholder as "tf.placeholder"?

File: efficientdet/third_party/graph_edit/examples/edit_graph_example.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+import tensorflow as tf
+import third_party.graph_edit as ge
+
+FLAGS = tf.flags.FLAGS
```
Comment (mingxingtan) [inline efficientdet/third_party/graph_edit/examples/edit_graph_example.py:25]: Looks like this FLAGS is not used anywhere. How about deleting this line?

File: efficientdet/third_party/graph_edit/tests/edit_test.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  """
+
+  def setUp(self):
+    """Set up."""
+    self.graph = ops.Graph()
+    with self.graph.as_default():
+      self.a = constant_op.constant([1., 1.], shape=[2], name="a")
+      with ops.name_scope("foo"):
+        self.b = constant_op.constant([2., 2.], shape=[2], name="b")
+        self.c = math_ops.add(self.a, self.b, name="c")
```
Comment (mingxingtan) [inline efficientdet/third_party/graph_edit/tests/edit_test.py:43]: If you simply import tensorflow, and use "tf.add" here; then it is much easier to read and maintain?

**Review by mingxingtan — COMMENTED** [review]
Looks looks good to me.

**Review by mingxingtan — COMMENTED** [review]
My biggest advice is to avoid importing tensorflow internal modules, which would make it much easier to read and maintain.

(internally, importing tensorflow internal modules is prohibited except very special cases.)

**Comment by mingxingtan** [comment]
I realized the original graph_edit is package of tf.contrib, which is an internal tf library, so it might be fine in this case.

I tried to run the tests under third_party/graph_edit, but they all failed. Could you help fixed those tests? Thanks!

File: efficientdet/third_party/graph_edit/__init__.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""TensorFlow Graph Editor."""
+
+from __future__ import absolute_import
```
[the remark this answers is not in the document]
Comment (NikZak) [reply efficientdet/third_party/graph_edit/__init__.py]: Thanks. Removed

File: efficientdet/third_party/graph_edit/examples/edit_graph_example.py (hunk 4 above)
Comment (NikZak) [reply efficientdet/third_party/graph_edit/examples/edit_graph_example.py]: Thanks. Removed

**Comment by NikZak** [comment]
The tests are running now. There were two special cases, testing tf.cond and tf.while in a copy of the graph, that did not pass. This is not needed for the purpose of efficientdet.

Please run the tests as described in the FAQ to avoid import errors
https://github.com/google/automl/blob/eb74c6739382e9444817d2ad97c4582dbe9a9020/efficientdet/g3doc/faq.md#22-how-can-i-run-all-tests

if you want to run with pytest make sure efficientdet folder is in PYTHONPATH

**Comment by mingxingtan** [comment]
@NikZak 
Is this PR stable now? If so, I will try to import this PR, but you need to freeze this PR (don't submit new commits to this branch).

**Comment by NikZak** [comment]
It's stable. OK, got it

File: efficientdet/third_party/grad_checkpoint/memory_saving_gradients.py [hunk 7]
```diff
@@ -0,0 +1,519 @@
+"""Recompute tensorflow graph reducing GPU memory usage."""
+import logging
+import contextlib
+import time
+import sys
+from toposort import toposort
+import numpy as np
```
Comment (mingxingtan) [inline efficientdet/third_party/grad_checkpoint/memory_saving_gradients.py:7]: Is it possible to replace this toposort with a function?  (We don't have this library internally).

Comment (NikZak) [reply efficientdet/third_party/grad_checkpoint/memory_saving_gradients.py:6]: Ok. Done

Comment (NikZak) [reply efficientdet/third_party/grad_checkpoint/memory_saving_gradients.py:6]: Ok. Done

**Comment by LucasSloan** [comment]
Why are we vendoring these dependencies?  It looks like the answer is that there were a bunch of tweaks to the underlying libraries?

**Comment by NikZak** [comment]
@LucasSloan
None of these libraries work with tf 2.x originally

Main changes:
Graph editing - minimal changes in the syntaxis, refactoring and some tweaks in the tests. 
Memory gradients - the algorithm of picking the gradients is substantially changed, the algorithm of gluing graph back together is left as is. Refactored. 
Nvgpu - a completely independent algorithm with same name as a pypi repo
Toposort - no changes, just refactoring

**Review by LucasSloan — APPROVED** [review]

**Comment by fsx950223** [comment]
What's difference between ```debug_print``` and ```logging.debug```.

File: efficientdet/third_party/grad_checkpoint/memory_saving_gradients.py [hunk 8]
```diff
@@ -0,0 +1,521 @@
...
+      logging.debug("ts_all: %s", len(ts_all))
+
+      # filter out all tensors that are inputs of the backward graph
+      with util.capture_ops() as bwd_ops:
+        tf_gradients(ys, xs, grad_ys, **kwargs)
+
+      bwd_inputs = [t for op in bwd_ops for t in op.inputs]
+      # list of tensors in forward graph that is in input to bwd graph
+      ts_filtered = list(set(bwd_inputs).intersection(ts_all))
+      debug_print("Using tensors %s", ts_filtered)
```
Comment (fsx950223) [inline efficientdet/third_party/grad_checkpoint/memory_saving_gradients.py:172]: What's difference between debug_print and logging.debug.
Maybe we should use absl.logging.
Could you add some test cases for memory_saving_gradients? Thanks.

Comment (NikZak) [reply efficientdet/third_party/grad_checkpoint/memory_saving_gradients.py:172]: Hi @fsx950223  
1) Regarding debug_print you can refer to the function description 
```python 
  """Like logger.log, but also replaces all TensorFlow ops/tensors with their
  names. Sensitive to value of DEBUG_LOGGING, see enable_debug/disable_debug
  Usage:
    debug_print("see tensors %s for %s", tensorlist, [1,2,3])
  """
```
If you want to log the tensors themselves debug_print is a good option (but it may be quite wordy). If no tensors standard logging.debug would do
2) Thanks for the suggestion. What does this library do? I do not want to add a dependency on any other third party library as it may lead to necessity to vendor them to this repo due to google internal rules. Then need refactoring and so on. Could be a neverending cycle 
3) Regarding testing. More testing coverage is always good. Probably I will add some tests at a later stage (maybe it should be part of a bigger integration test for this repo as gradient checkpointing is a major switch) and probably I will add some optimizations to the algorithm. However, this pull request is stable and it is on the way to be merged so unless it is a bug I don't want to make any changes

Comment (fsx950223) [reply efficientdet/third_party/grad_checkpoint/memory_saving_gradients.py:172]: We always use absl.logging in the repo.

Comment (NikZak) [reply efficientdet/third_party/grad_checkpoint/memory_saving_gradients.py:172]: Thanks, as it is not a bug, I will rectify after merging this pull request

**Comment by mingxingtan** [comment]
@NikZak @fsx950223 @LucasSloan 
I am merging this PR now, followed by another PR to fix some format issues.
